src/com/ms/silverking/cloud/dht/daemon/storage/convergence/management/ComputeRingStats.java

```diff
@@ -29,215 +29,216 @@
+
+    lostServers = generateLostServers(s);
+    for (Set<IPAndPort> replicaSet : this.sMap.getReplicaSets()) {
+      boolean lost;
+
+      lost = true;
+      for (IPAndPort replica : replicaSet) {
+        if (!lostServers.contains(replica)) {
+          lost = false;
+          break;
         }
-        return (double)losses / (double)simulations;
+      }
+      if (lost) {
+        return true;
+      }
     }
-    
-    private boolean lossTrial(int s) {
-        Set<IPAndPort>    lostServers;
-        
-        lostServers = generateLostServers(s);
-        for (Set<IPAndPort> replicaSet : this.sMap.getReplicaSets()) {
-            boolean    lost;
-            
-            lost = true;
-            for (IPAndPort replica : replicaSet) {
-                if (!lostServers.contains(replica)) {
-                    lost = false;
-                    break;
-                }
-            }
-            if (lost) {
-                return true;
-            }
-        }
-        return false;
+    return false;
+  }
+
+  private Set<IPAndPort> generateLostServers(int s) {
+    Set<IPAndPort> lostServers;
+
+    // Currently a simple algorithm appropriate only when s <<< |replicas|
+    lostServers = new HashSet<>();
+    while (lostServers.size() < s) {
+      int index;
+      IPAndPort replica;
+
+      index = ThreadLocalRandom.current().nextInt(replicas.size());
+      replica = replicas.get(index);
+      if (!lostServers.contains(replica)) {
+        lostServers.add(replica);
+      }
     }
-    
-    private Set<IPAndPort> generateLostServers(int s) {
-        Set<IPAndPort>    lostServers;
-        
-        // Currently a simple algorithm appropriate only when s <<< |replicas|
-        lostServers = new HashSet<>();
-        while (lostServers.size() < s) {
-            int            index;
-            IPAndPort    replica;
-            
-            index = ThreadLocalRandom.current().nextInt(replicas.size());
-            replica = replicas.get(index);
-            if (!lostServers.contains(replica)) {
-                lostServers.add(replica);
-            }
-        }
-        return lostServers;
+    return lostServers;
+  }
+
+  public void computeLossProbability(int[] serversLost, int simulations) {
+    out.println("*******************************");
+    for (int s : serversLost) {
+      double lp;
+
+      lp = computeLossProbability(s, simulations);
+      out.printf("%d\t%.7f\n", s, lp);
     }
-    
-    public void computeLossProbability(int[] serversLost, int simulations) {
-        out.println("*******************************");
-        for (int s : serversLost) {
-            double    lp;
-            
-            lp = computeLossProbability(s, simulations);
-            out.printf("%d\t%.7f\n", s, lp);
+  }
+
+  ///////////////////////////////////////////////////////////
+
+  public boolean lostData(File lostServerFile) throws IOException {
+    return lostData(new FileInputStream(lostServerFile));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when ComputeRingStats.lostData throws an exception. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).